### ISSUE TITLE
emxomf: Out of memory

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/libc#82]
Faced another issue while building Chromium (https://github.com/bitwiseworks/qtwebengine-chromium-os2/issues/3#issuecomment-636993816). EMXOMF hangs the system hard (cold reboot needed) when several instances are converting really big (> 30 MB) AOUT files into OMF in parallel in SMP mode. Reducing the number of parallel executions makes it not hang but spit:
```
emxomf: Out of memory
```
sometimes instead. In UNI mode it seems to work fine but really slowly.

I've tried to look at memory usage by Theseus right before it's about to hang but unfortunately,
Theseus just crashes when I try to query EMXOMF process details under heavy load from the real Chromium build environment (ninja).

I'm now trying to reproduce the conditions by compiling 4 several huge (> 30 MB) ASM files in parallel with `gcc -Zomf -c` (it runs both AS and EMXOMF). In this test scenario I can't make it hang though yet and Theseus also works (though hardly operable because of 100% CPU load and PM non-responsiveness). I see this so far in the RAM usage by process output:
```
RAM Usage by Process:
--------- private --------   ------ owned shared ------
   bytes   Kbytes   Mbytes      bytes   Kbytes   Mbytes   who
0B56E000   185784   181.430   00310000     3136    3.063   system
00002000        8    0.008                                Sysinit
00086000      536    0.523   00001000        4    0.004   SSESCTL
0000D000       52    0.051   00002000        8    0.008   CACHEF32
00005000       20    0.020   00001000        4    0.004   LVMALERT
00000000        0    0.000                                LANMSGEX
00005000       20    0.020   00001000        4    0.004   CNTRL
00063000      396    0.387   00021000      132    0.129   DHCPCD
00000000        0    0.000                                DOSCTL
00000000        0    0.000                                MIDIDMON
00000000        0    0.000                                MMHELPDD
00010000       64    0.063   00005000       20    0.020   ACPIDAEM
0001F000      124    0.121   00764000     7568    7.391   PMSHELL
00004000       16    0.016                                VDM
00000000        0    0.000                                HARDERR
00000000        0    0.000   00003000       12    0.012   PMSPOOL
000F7000      988    0.965   005D7000     5980    5.840   PMSHELL
0000A000       40    0.039   00002000        8    0.008   TWOHAND
00031000      196    0.191   00040000      256    0.250   XWPDAEMN
00000000        0    0.000                                USBMON
00009000       36    0.035   00006000       24    0.023   SOCKTIDY
00087000      540    0.527   00033000      204    0.199   NDCTL
00000000        0    0.000                                CAD_POP
00000000        0    0.000   00091000      580    0.566   CUPSD
00000000        0    0.000                                SSHD
0003C000      240    0.234   00BEB000    12204   11.918   VNCSERVE
00043000      268    0.262   0014B000     1324    1.293   SMBD
000B6000      728    0.711   00131000     1220    1.191   NMBD
00001000        4    0.004                                NMBD
0000B000       44    0.043   00035000      212    0.207   SMBD
000C3000      780    0.762   00004000       16    0.016   FC
0001A000      104    0.102   00007000       28    0.027   DAYTIME
00255000     2388    2.332   0000B000       44    0.043   KTASKMGR
0003D000      244    0.238   0017B000     1516    1.480   SL
00079000      484    0.473   00004000       16    0.016   FC
001D9000     1892    1.848   00004000       16    0.016   FC
00063000      396    0.387   00141000     1284    1.254   SMBD
00063000      396    0.387   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005A000      360    0.352   00141000     1284    1.254   SMBD
0005B000      364    0.355   00141000     1284    1.254   SMBD
00210000     2112    2.063   00141000     1284    1.254   SMBD
01A40000    26880   26.250   00033000      204    0.199   THESEUS4
00037000      220    0.215   00098000      608    0.594   DASH
00040000      256    0.250   00055000      340    0.332   XGCC
00037000      220    0.215   00098000      608    0.594   DASH
00046000      280    0.273   05BCE000    94008   91.805   AS-NEW
00040000      256    0.250   0004A000      296    0.289   XGCC
00037000      220    0.215   00098000      608    0.594   DASH
00040000      256    0.250   0004A000      296    0.289   XGCC
00048000      288    0.281   00098000      608    0.594   DASH
00026000      152    0.148   19907000   418844   409.027   EMXOMF
00046000      280    0.273   053A2000    85640   83.633   AS-NEW
0008A000      552    0.539   0004A000      296    0.289   XGCC
00046000      280    0.273   07344000   118032   115.266   AS-NEW
00088000      544    0.531   0720E000   116792   114.055   AS-NEW
00026000      152    0.148   0D842000   221448   216.258   EMXOMF
00026000      152    0.148   11CC6000   291608   284.773   EMXOMF
0003B000      236    0.230   13B86000   323096   315.523   EMXOMF
--------   ------   ------   --------   ------   ------
0E4F4000   234448   228.953   696A6000   1727128   1686.648   total RAM in use
540C2000   1377032   1344.758   free RAM
--------   ------   ------
CBC5C000   3338608   3260.359   total of all RAM pages found (Pvt + Shr + Free)
< End of THESEUS4 (v 4.001.00) output @ 15:13:59 on 3.6.2020 >
```
EMXOMF memory usage is quite impressive. But apparently not enough to trigger the crash.

**Comment by dmik** [comment]
Ok, feeding it 6 EMXOMF processes in parallel made it hang like hell. And thanks to the VNC session I had, here is the postmortem screenshot:
![Снимок экрана 2020-06-03 в 14 58 18](https://user-images.githubusercontent.com/900726/83637852-ebf8ce00-a5b0-11ea-9f60-9550bca3a53e.png)

As one may see, my assumption about OOM is 100% correct. Six EMXOMF processes have eaten all the available memory. With one of them eating as much as 1 GB. No big surprise Kernel goes panic under such a stress. It's strange why swap file didn't help here. Perhaps because it's just 2MB :) Is it possible to set it to something like 2 GB? I'm not sure the OS/2 Kernel can handle this...

Another question I have is why EMXOMF is using shared memory like hell... Makes no sense to me. I need to check its memory allocation bits. May be it will give me some hints on how to make succeed under such stress. Any ideas are welcome.

**Comment by dmik** [comment]
Note that I have VIRTUALADDRESSLIMIT=1536 now but this doesn't help. I also tried VAL=3072 — the result was just the same (but may be later, I didn't measure).

**Comment by dmik** [comment]
EMXOMF allocates memory with a simple `malloc` call which uses high memory due to `-Zhigh-mem` in its build options. So I guess that Theseus4 knows nothing about high memory and misinterprets high memory blocks as shared memory then. I have an indirect proof of that by requesting a per-process Memory Utilization report. While the amount of "shared" memory of a running EMXOMF instance continues to grow as it processes the AOUT file, the amount of total allocated memory for this process in the Memory Utilization report stays still (at 52 M). My guess is that this report only sees low memory and therefore doesn't see high-memory allocation actually taking place. So, in general, Theseus is not very useful in this scenario...

**Comment by dmik** [comment]
I also tried `SWAPPATH=C:\OS2\SYSTEM 1048576 1048576` (and it actually worked by letting me have an 1G swap file) but this didn't have any effect on the case. It still runs out of memory. So the only solution here I see is to optimize EMXOMF by dramatically reducing simultaneous memory allocations (perhaps involving hard disk to store intermediate results or such, by discarding excessive debug info, etc.). But I'm afraid this may result in a full rewrite of the tool. I don't feel like I have time for that task now. And BTW, even if/when I (or someone else) have, it looks more useful to invest this time in ELF support (see https://github.com/bitwiseworks/gcc-os2/issues/12).

I will postpone it for now by lowering simultaneous processes with -j X. If I stumble upon it again, I might come up with some quick hacks as well.

**Comment by dryeo** [comment]
The swap file should dynamically grow up to 2GBs or so, where the system 
will crash with a swapfile full error.
I think the problem is more a shortage of address space where the swap 
file will only help if you don't have enough real ram.

**Comment by dmik** [comment]
One more observation. Disabling handling debug info (.stab* records) with just an `-s` option to `emxomf.exe` when converting from AOUT to OMF not only makes conversion a blink instead of several minutes, it also solves the "out of memory" problem completely.

So, I will quickly try to reduce the amount of debug info in some meaningful way but if I fail I will just strip it off completely. Debug info is useful for .TRP file generation when we get crashes and such, but in the absence of object files we won't build executables and they won't crash :) So it's still a way to go.

**Comment by dmik** [comment]
I've worked around it by making the `convert_debug` (https://github.com/bitwiseworks/libc/blob/d9319ee594eb12ee409f0f284bbd9f26d2abc352/src/emx/src/emxomf/stabshll.c#L3939) a no-op.

Now converting that giant  55 MB AOUT file to OMF happens in a blink.

I've also tried using a test hello.exe app built in `-g` mode with this EMXOMF in the debugger and passing it through a trap to generate a .TRP. From what I see making `convert_debug` a no-op only disables local variable processing and hides static function names from HLL (.TRP doesn't seem to be able to display static functions anyway but that perhaps a bug-o-feature of EXCEPTQ, or EMXOMF just does it wrong). So this is the "normal" stack trace from .TRP:
```
   EBP     Address    Module     Obj:Offset    Nearest Public Symbol
 --------  ---------  --------  -------------  -----------------------
 Trap  ->  00010050   HELLO     0001:00000050  hello.cpp#19 _main - 5D 0001:000000AD (hello.cpp)

 0012FF18  000100E1   HELLO     0001:000000E1  hello.cpp#29 main + 34 0001:000000AD (hello.cpp)

  Offset Name                 Type                         Hex Value
  ÄÄÄÄÄÄ ÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄ ÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄ ÄÄÄÄÄÄÄÄÄ
   8     argc                 32 bit signed                1
   12    argv                 pointer to type 0x203        12FF94
   20    ppib                 pointer to type 0x206        12FF64
   28    s                    pointer to 8 bit unsigned    12FF8C
   24    i                    32 bit signed                1E16BB85

 0012FF48  00010027   HELLO     0001:00000027  crt0.s#90 __text + 27 0001:00000000 (D:\Temp\cckbIdJn.s)

 0012FF74  1E16C031   LIBCX0    0001:0000C031   ___init_app + 11 0001:0000C020 (main.obj)

 0012FFE0  1DF244BB   LIBCN0    0001:000344BB  appinit.s#16 ___init_app + B 0001:000344B0 (appinit.obj)
```
And this is the stack trace with `convert_debug` disabled:
```
   EBP     Address    Module     Obj:Offset    Nearest Public Symbol
 --------  ---------  --------  -------------  -----------------------
 Trap  ->  00010050   HELLO     0001:00000050  hello.cpp#19 _main - 5D 0001:000000AD (hello.cpp)

 0012FF18  000100E1   HELLO     0001:000000E1  hello.cpp#29 main + 34 0001:000000AD (hello.cpp)

 0012FF48  00010027   HELLO     0001:00000027  crt0.s#90 __text + 27 0001:00000000 (D:\Temp\cckbIdJn.s)

 0012FF74  1E16C031   LIBCX0    0001:0000C031   ___init_app + 11 0001:0000C020 (main.obj)

 0012FFE0  1DF244BB   LIBCN0    0001:000344BB  appinit.s#16 ___init_app + B 0001:000344B0 (appinit.obj)
```
Other than that, .TRP files are identical. This said, source file line information *IS* preserved as well as public function names. I also tried IBM Debugger with the test EXE — same story. Source level debugging is in place, just local variable (and function parameter) information is missing.

But variable and static function debugging has never worked well with EMXOMF (its HLL support is highly experimental). Needless to say that IBM debugger is very unstable because it's too old and limited function wise so I use it very rarely in my debugging sessions. So the above result is more than enough for debugging purposes I must say.

I will try the hacked EMXOMF "in production" with Chromium for some time before a final decision but it looks promising as a temporary (=long term, hehe) solution.

**Comment by dmik** [comment]
Re problems with debugging variables, see e.g. https://github.com/bitwiseworks/gcc-os2/issues/5.

**Comment by StevenLevine** [comment]
Sorry for the late reply.  My front door decided to lock itself permanently.  It's partially fixed which is sufficient for the moment, but I've started looking a new locksets.  I guess they are designed to wear out after 31 years.  Of the 4 condos in my complex, only one will be left with the original lockset.

Anyway, back on topic...

When looking a memory, it's often best to start with System->Linear Usage by Process.  If the window fills, up use my LinearMemoryMap.cmd from www.warpcave.com/os2diags.  This will show EMXOMF's private memory usage and will show you if this usage is sufficient to cause problems with lower shared memory.  Running out of lower shared memory is often the cause of system hangs.

The problem with EMXOMF is not so much that it using a lot of memory, but that Theseus claims it is using a lot of shared memory:

--------- private --------   ------ owned shared ------
   bytes   Kbytes   Mbytes      bytes   Kbytes   Mbytes   who
00026000      152    0.148   19907000   418844   409.027   EMXOMF

You might be right about Theseus reporting the memory incorrectly, but it does understand upper private memory.  You can see this in the Linear Usage by Process report.

As Dave noted the swap file can theoretically grow to 2GB.  The SWAPPATH paramaters set the warning limit and the initial size.

I'm not sure you are running out of address space.  Each process gets VIRTUALADDRESSLIMIT of address space.  What is more likely is you are running out of memory objects.  We are limited to 65K memory objects.

Regarding your tweaks to convert_debug, I generally don't make use of the local variable value reporting, so it's not much of a loss.  The file and line number reporting is what makes the debug data useful to me.

I am not sure why the static functions are not reported, but I just checked and I can confirm that exceptq is getting it wrong.  Probably just some missing code.  It gets the file and line# correct.  I will look at this.

Regardling missing local variables, I've noticed that if I build exceptq with /O and icc and debug with idebug, local variables are gone.  I attributed this to a limitation/defect in idebug.

Regarding idebug stability, it crashes here too, but not so much to make it unusable.  To workaround the trashed *.@oa files, I have scripts to backup and restore them as needed.  I find idebug works better if I try to avoid running an application to completion.  I shut down idebug and let it terminate the app.

**Comment by dmik** [comment]
Nice door lock story :) (I guess 31 years have already passed).

I will check the Linear Usage by Process report but regarding the RAM Usage by Process report, I'm sure Theseus misinterprets high private memory as shared mem. Because EMXOMF only uses LIBC malloc and nothing more and the default heap does never go shared mem (but it goes high private arena if it's available).

I think I'm seeing both scenarios here — running out of memory objects (when too many EMXOMF instances are working in paralell) and running out of address space (when once instance is converting an enormously large AOUT file). What you see on the screenshot, is the first case. But when I was playing with the large AOUT file, I saw EMXOMF alone eating almost 2G of "shared" memory alone and that was very close to thee VAL limit. Anyway, the `convert_debug` tweak should make both cases much less likely.

Re the debugger, I more mean its functional limitations than its (in)stability. Modern debuggers provide various kinds of configurable "fences" that let kick in to the code at certain moments. As well as powerful evaluation engine for dealing with c/c++ variables and such. We lack all that so in complex scenarios I have to manually inject "printf"-like stuff to understand what's going on.

**Comment by dmik** [comment]
As https://github.com/bitwiseworks/libc/issues/72#issuecomment-644955857 and below comments show, there is a side effect of the above hack with the debug info leading to crashes of WL.EXE when building some GCC parts. EMXOMF.EXE needs to be rolled back to the original to build it. And then again to he hacked version to build Chromium.

@StevenLevine will try to fix WL.EXE in the mean time.

**Comment by StevenLevine** [comment]
I have some info on the wlink trap.  It appears that the relocation that
is causing the problem is in libiberty's hex.obj.  There is either
something wrong with OpenWatcom's understanding of hex.obj or hex.obj
itself if subtly corrupted.  What I did was extract hex.obj and
xstrerror.obj from the library.  xstrerror is my control group. 
OpenWatcom's dmpobj seems to understand both object files object file well
enough to produce what looks like valid output.  However wdis claims that
something is wrong with hex.obj and wdis has no problems with xstrerror.h.

Please rebuild hex.o and libiberty.a and let me know if this resolves the
wlink exception.

Before rebuilding, it might be interesting to extract hex.o from
libibery.a and see if objdump complains about the .o file.

I will continue testing with wdis since it should be easier to see what it
causing wdis to complain.  Even if hex.obj is really corrupted, I would
prefer to add some code to avoid the trap.

**Comment by StevenLevine** [comment]
After a refresher course on the OMF format, I now have a better
understanding what is broken and how.

For as yet known reasons, hex.obj is broken.  Dmpobj reports

FIXUPP386(9d) recnum:17, offset:000002c1h, len:0009h, chksum:b6h(b6)
    0006h Seg   BASE         Frame: THREAD 7  Target: SI(1)
    0008h Seg   OFFSET386    Frame: THREAD 7  Target: SI(1)

This record is missing the required Thread subrecord.  Threads must be
defined before they can be used.  Compare this to xstrerror.obj which
reports:

FIXUPP386(9d) recnum:16, offset:0000014fh, len:0018h, chksum:03h(03)
    Thread: FRAME  0: GI(1)   
    0025h Self  OFFSET386    Frame: THREAD 0  Target: EI(1)
    Thread: TARGET 0: SI(1)
    003bh Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 0
    Thread: TARGET 1: SI(3)
    0042h Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 1
    0047h Self  OFFSET386    Frame: THREAD 0  Target: EI(2)
    004ch Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 1

The THREAD subrecords are indexed by the Thread numbers used in the FIXUP
subrecords.

The malformed record is not a problem for dmpobj because it is only
listing the record content.  Both wdis and wlink have a problem with
hex.obj because it is malformed.

What is currently unknown is if gcc generated a malformed hex.o or if
emxomf messed up when converting hex.o to hex.obj.

I will update wdis to report the error more clearly and wlink to fail
rather than trapping.

**Comment by dmik** [comment]
Re `hex.obj`. In the `libibertly.a` file that you have in the test case, this file is in A.OUT format (under `hex.o` name). When you the test script, `emxomfld.exe` detects that `libiberty.a` is a Linux library with A.OUT files, and converts it to an OS/2 library with OMF files using the supplied `emxomf.exe` (that contains my hack). Then this library is passed on to WL and it crashes. If you delete `emxomf.exe` so that a non-hacked version from RPM is used, the WL crash is gone. So, there is nothing for me to rebuild. All involved things happen on your side.

The question we need to answer here is who is guilty: my hacked `emxomf.exe` that converts A.OUT `hex.o` to OMF `hex.obj` so that WL crashes, or WL itself which just doesn't expect an unusual (but pretty legal) `hex.obj` contents.

**Comment by StevenLevine** [comment]
I understand the processing that is done.

Based on what you are saying here it would appear that your hacked
exmomf.exe generates the broken hex.obj.  The fixup subrecord contans
thread references which means it must define a thread subrecord.  If you
wish, I can point you to the relevant OMF specs.

wlink is wrong in the sense that it should have detected and reported the
error rather than crashing.  I have fixed this locally.

wdis detects the error and reports it, although not all that clearly.  The
error message is a generic catchall.

VAC's icc, does not like hex.obj either:

ILink : warning LNK4021: no stack segment
D:\sla_dev2\TestVACVIO\0\hex.obj(hex.obj) : warning LNK4008: aliased
fix-up to non-alias object near 146 in object TEXT32
D:\sla_dev2\TestVACVIO\0\hex.obj(hex.obj) : warning LNK4008: aliased
fix-up to non-alias object near 146 in object TEXT32
D:\sla_dev2\TestVACVIO\0\hex.obj(hex.obj) : warning LNK4001:
frame-relative fix-up, frame ignored near 0146 in segment TEXT32
D:\sla_dev2\TestVACVIO\0\hex.obj(hex.obj) : warning LNK4001:
frame-relative fix-up, frame ignored near 0146 in segment TEXT32
D:\sla_dev2\TestVACVIO\0\hex.obj(hex.obj) : warning LNK4001:
frame-relative fix-up, frame ignored near 0148 in segment TEXT32
D:\sla_dev2\TestVACVIO\0\hex.obj(hex.obj) : warning LNK4001:
frame-relative fix-up, frame ignored near 0148 in segment TEXT32

If you modify emxomf again, I recommend you use wdis to check the .obj
file.  It does more extensive checking of the file content that wlink
appears to.

**Comment by StevenLevine** [comment]
wl_exe-20200622.zip has been uploaded to www.warpcave.com/openwatcom.  This will avoid the trap we encountered from hex.obj.  Note that wl.sym is Dwarf debug data, not mapsym output.

**Comment by dmik** [comment]
Thanks, Steve! Will test it ASAP (doing some work with #84 now).

**Comment by StevenLevine** [comment]
I now know what your modified emxomf did to break the conversion of hex.o to hex.obj.  The unmodified emxomf generates the following fixup record:

FIXUPP386(9d) recnum:18, offset:000002f0h, len:0017h, chksum:ddh(dd)
    Thread: FRAME  0: GI(1)   
    Thread: TARGET 0: SI(1)
    0020h Seg   BASE         Frame: THREAD 0  Target: THREAD 0
    0027h Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 0
    0046h Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 0
    0052h Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 0
    0067h Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 0
    006bh Seg   BASE         Frame: THREAD 0  Target: THREAD 0

This record is missing from the hex.obj generated by your modified emxomf.

I've not looked at your modifications is sufficient detail to know why this occurred.  However at a minimum, the thread FRAME definition must be retained.

**Comment by StevenLevine** [comment]
emxomf -s produces a valid .obj file.

**Comment by dmik** [comment]
Yes, -s drops ALL the debug info. I found that out long ago but I'm pretty sure I will desperately need that debug info when trying to run it. So I will have to dig into EMXOMF internals and find out how to emit FRAME w/o emitting the rest.

**Comment by StevenLevine** [comment]
In this case you actually want to omit the fixup records.  If you look at
what they reference you will see that they refer to debug related data
that was stripped by your modified emxomf.  The unmodified emxomf with -s
correctly stripped this debug data and the related fixup records.

Here are the relevant OMF records produced by the unmodified emxomf:

LEDATA(a0) recnum:17, offset:0000026fh, len:007eh, chksum:1fh(1f)
    Seg index:4 offset:00000000h
    00000000 80|1b 40|02 04|20 20|20 20|0b 4d|61 72|20 32|38 <..@..   
.Mar 28>
    00000010 20|32 30|32 30|10 1e|31 00|16 06|e4 07|80 05|11 <
2020..1........>
    00000020 00|00 00|00 80|1d 1d|00 00|00 00|00 02|40 01|00
<.............@..>
    00000030 00|00 00|40 01|00 00|00 00|08 08|68 65|78 5f|69
<...@.......hex_i>
    00000040 6e|69 74|80 0a|00 00|00 00|00 40|00 00|00 00|80
<nit.......@.....>
    00000050 0a|00 00|00 00|00 40|00 00|00 00|80 01|02 80|01
<......@.........>
    00000060 02|80 01|02 80|14 05|40 00|00 00|00 00|01 02|0a
<.......@........>
    00000070 5f|68 65|78 5f|76 61|6c 75|65                   <_hex_value>

FIXUPP386(9d) recnum:18, offset:000002f0h, len:0017h, chksum:ddh(dd)
    Thread: FRAME  0: GI(1)   
    Thread: TARGET 0: SI(1)
    0020h Seg   BASE         Frame: THREAD 0  Target: THREAD 0
    0027h Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 0
    0046h Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 0
    0052h Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 0
    0067h Seg   OFFSET386    Frame: THREAD 0  Target: THREAD 0
    006bh Seg   BASE         Frame: THREAD 0  Target: THREAD 0

FIXUPP386(9d) recnum:21, offset:000003a5h, len:0009h, chksum:96h(96)
    0006h Seg   BASE         Frame: THREAD 0  Target: SI(1)
    0008h Seg   OFFSET386    Frame: THREAD 0  Target: SI(1)

The threads referenced in the last fixup record are defined in a prior
fixup record which is perfectly legal.  Index 4 refers to the BSS32 data
which appears to be generated by emxomf as part of generating debug data.

With emxomf -s, these 3 records are not generated.

With your modified emxomf, the last record is still generated.  This is
needs to be fixed.

Regarding the size of debug data, you can save a lot of space by stripping
the path prefix.  For example, you have:

LINNUM386(95) recnum:19, offset:0000030ah, len:0076h, chksum:6fh(6f)
    File names table
    Record Number of Start of Source:      0
    Number of Primary Source Records:      0
      Source file      1:
'D:\Coding\gcc\master-build\build-i686-pc-os2-emx\libiberty\..\..\..\master\libiberty\hex.c'

This can be written to hex.c.  Idebug can still find the source files if
you set PMDPATH.

Another option, which I have recommended before back in the days of your
Mozilla work is to enhance the build system to efficiently generate debug
data for a subset of the code.  The reality is you will never need to
debug a large percentage of the code.  It will just work.

One way to do this is to enhance emxomf to take a list of wildcarded *.o
pathnames which specifies the files to have -s applied to.

**Comment by dmik** [comment]
Okay, I see. Thanks for the ideas. I will think on them. Though I don't really like the idea of stripping paths. It is VERY convenient. PMDPATH is good but if you have hundreds of source directories (like in Chromium) this is a NO way to go. Also, selectively applying c++ debug info by a wildcard doesn't seem as a workable solution because of templates, weak symbols, virtual functions etc. I guess I will easily end up with simply incomplete debug info this way. But there is something to think on.

**Comment by dmik** [comment]
What could be done though is **normalizing** paths. It will save some space (I really wonder why emx/wl doesn't do that, it makes no sense).

**Comment by StevenLevine** [comment]
>What could be done though is **normalizing** paths. It will save some
>space (I really wonder why emx/wl doesn't do that, it makes no sense

Sure it does.  You just need a sense of history.  When EMX was invented
folks still tended to use short file names because they were coming out of
the DOS world.  Even Unix tended to use short file names because it was
invented in the days of the ASR-33 teletype and targeted systems such as
PDP-8 and PDP-11's.  We all wanted to minimize the number of characters we
needed to type.

Much the same is true for OpenWatcom.  Wikipedia tells us that the first
version was for the S/370.  The 370's I worked on maxed out a 256K of RAM. 
The next Watcom target was DOS and things just grew from there.

When I started, Intel was not even making CPU chips.  CPU chips basically
did not exist.  I've built production software that shipped on early CPU
chips such as 8085's and 8080's.  These all maxed out at 64KB of memory,
but we rarely had that much in a system.

I am not going to fault the implementors of EMX and Watcom for not
planning for today's 64-bit systems with 32GB of RAM and applications the
size of a QT5 based browser.  The certainly had to way to test these kinds
of limits.

The thought of implementing something like QT5 in the 80's and 90's was
interesting to think about, but not something that was a day to day
concern.

Normalizing paths is a good idea. It will save some space.  I would also
give some consideration to converting stored paths to relative paths. 
Let's say you stripped

D:\Coding\gcc\master-build\build-i686-pc-os2-emx\

from a normalized:

D:\Coding\gcc\master-build\build-i686-pc-os2-emx\libiberty\..\..\..\master\libiberty\hex.c

which a relatively small number PMDFPATH entries, you will significantly
reduce the space needed to store paths in the debug data.  What percentage
of the total space required for debug data this will save, I can only
speculate until I review the HLL specs.  It's been a while since I added
HLL support to the linker so the details are fuzzy.

Even with this kind of reduction the debug data is going to require a lot
of space to define local variables and line numbers and such.

That's said, I am willing to look at adding path normalization to the
debug data processing.  It should not be that difficult.  It will be less
work than real_path and friends because the paths are known to be be valid
on input.

More importantly, I need to understand why you are getting the OOM
failure.  OpenWatcom implements both a virtual memory scheme and a spill
file scheme to handle relocation data and debug data.  I've not needed to
know the details of these components, so for now, I don't.  It's possible
something is going wrong that allow too much address space to be used.

What I need to see is a failing link run with OPTION VERBOSE enabled so
that I have some idea of when the linker is falling over.

Also, you might want to try OPTION NOCACHE and still if this reduces
memory utilization.  Wlink implments caching in 3 states.  The linker
options are NOCACHE and CACHE.  If neither are specified, the linker will
cache what the docs call "small libraries."  However, with enough of
these, I guess it is possible to run out of memory.

Steven

-- 
----------------------------------------------------------------------
"Steven Levine" <steve53@earthlink.net>  Warp/DIY/BlueLion etc.
www.scoug.com www.arcanoae.com www.warpcave.com
----------------------------------------------------------------------

**Comment by StevenLevine** [comment]
Turns out normalizing the paths in wlink is non-starter.  I conveniently
forgot that the debug data itself is opaque by the time it gets to the
linker.

**Comment by dmik** [comment]
Note that debug Chromium builds started to fail at simultaneous compilation badly due to OOM for some reason so I had to limit the number of ninja jobs from the default 3 to 2 (using `export NINJAJOBS=-j2` which is understood by qmake generated Makefiles and passed down to NINJA.EXE). This helped. But it of course slows down the build.

Note that release builds seem to go fine with the default of -j3. I don't clearly recall any OOM hangs (but I could see it once or twice though).

It's not necessarily EMXOMF, it might be GCC (CC1PLUS.EXE) as well which eats quite a bit of memory too (more than 1 GB per a compilation unit is not rare).

Ideally, we should fix it in all three places:

1. OS/2 Kernel — fix swapping when there is enough of address space in each process, but just not enough physical memory. I don't have any stable proofs that it's often (or ever) the case, but numbers that I see in THESEUS suggest that it might be so. In either case, Kernel should NOT crash and hang the whole system when it is either out of address space in a given process or out of physical memory and out of swap space. It should just terminate the process with SYS0008 or such. Frequent hangs are really annoying and slow down compilation considerably because it requires a reboot, environment restoring and such. Having it make the process crash in a controlled way would make it possible to restart the compilation so that it eventually finishes (which is crucial for overnight and other unattended builds which otherwise just hang and remain such until rebooted by hand).

2. Limit GCC to never allocate more than, let's say 1G or 1.5G to make sure the process does not run out of address space. I have no idea if GCC has anything like that built in (and if it implements some sort of application side swapping when it needs more) or we need to reinvent it from scratch.

3. Do the same as above to EMXOMF (and to WL, see #84).

This is the only way to completely solve the current problems with building Chromium (and keep using parallelism in NINJA up to the real number of CPUs which is 4 in case of most systems). Neither of these three tasks are easy though.

**Comment by dmik** [comment]
Regarding GCC, there is actually `--param ggc-min-expand` and `ggc-min-heapsize`, see https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#Optimize-Options. The docs say:
```
ggc-min-expand

    GCC uses a garbage collector to manage its own memory allocation. This parameter specifies the minimum percentage by which the garbage collector’s heap should be allowed to expand between collections. Tuning this may improve compilation speed; it has no effect on code generation.

    The default is 30% + 70% * (RAM/1GB) with an upper bound of 100% when RAM >= 1GB. If getrlimit is available, the notion of “RAM” is the smallest of actual RAM and RLIMIT_DATA or RLIMIT_AS. If GCC is not able to calculate RAM on a particular platform, the lower bound of 30% is used. Setting this parameter and ggc-min-heapsize to zero causes a full collection to occur at every opportunity. This is extremely slow, but can be useful for debugging.
ggc-min-heapsize

    Minimum size of the garbage collector’s heap before it begins bothering to collect garbage. The first collection occurs after the heap expands by ggc-min-expand% beyond ggc-min-heapsize. Again, tuning this may improve compilation speed, and has no effect on code generation.

    The default is the smaller of RAM/8, RLIMIT_RSS, or a limit that tries to ensure that RLIMIT_DATA or RLIMIT_AS are not exceeded, but with a lower bound of 4096 (four megabytes) and an upper bound of 131072 (128 megabytes). If GCC is not able to calculate RAM on a particular platform, the lower bound is used. Setting this parameter very large effectively disables garbage collection. Setting this parameter and ggc-min-expand to zero causes a full collection to occur at every opportunity.
```
I need to try those.

**Comment by dmik** [comment]
Options seems to work but of course this caused a full rebuild of 10k compilation units (as these options are used with each of them). I will leave it overnight in -j4 mode (i.e. 4 compilation jobs at a time) to see if it survives this way.

**Comment by dmik** [comment]
Well, these options obviously don't help. First, the system died with zero free mem (I had a Theseus running) when there were 4 CC1PLUS processes each eating around 700 MB of memory (and NINJA eating around 200 MB). This is certainly a guilt of the kernel as it's clearly enough address space for all 5 processes. It should have started using the swap file but it didn't happen. A system halt happened instead.

Second, There is one file, `browser/browser/render_frame_host_impl.c`. It's 270 KB in size and the generated OMF object file is 730 MB (!!). When compiling, CC1PLUS consumes around 1.5 GB of system memory in total. Playing with the above options didn't have any effect on the maximum heap size. So, apparently for a source file that big GCC needs a large amount of memory which cannot be garbage collected to perform all required code analysis and generation. And I don't see any other options to limit its memory usage. Apparently, the only way to go here is to reduce the number of jobs to 2 or even to 1...

**Comment by dmik** [comment]
JFYI - -j2 worked well, no OOMs until the end of the build. We will have to stick with that until the kernel is fixed.
